File: persistence/src/main/java/eu/mico/platform/persistence/model/Content.java
package eu.mico.platform.persistence.model;

import org.apache.commons.vfs2.FileSystemException;
import org.openrdf.model.URI;

import java.io.InputStream;
import java.io.OutputStream;
import java.util.UUID;

/**
 * Binary content, offered through inputstreams and outputstreams. Can be used for representing any kind of non-RDF
 * content.
 *
 * @author Sebastian Schaffert (sschaffert@apache.org)
 */
public interface Content {

    /**
     *  Return the URI uniquely identifying this content part. The URI should be either a UUID or constructed in a way
     *  that it derives from the ContentItem this part belongs to.
     * @return
     */
    public URI getURI();

    /**
     * Return a new output stream for writing to the content. Any existing content will be overwritten.
     * @return
     */
    public OutputStream getOutputStream() throws FileSystemException;

    /**
     *  Return a new input stream for reading the content.
     * @return
     */
    public InputStream getInputStream() throws FileSystemException;
}

File: persistence/src/main/c++/Content.cpp
#include <boost/algorithm/string.hpp>

#include "Content.hpp"

using namespace std;
using namespace boost;

namespace mico {
  namespace persistence {

    Content::Content(const string baseUrl, const URI& uri) : baseUrl(baseUrl) {
      if(!starts_with(uri.stringValue(),baseUrl)) {
	throw string("the baseUrl is not a prefix of the URI, invalid argument");
      }

      id = uri.stringValue().substr(baseUrl.length() + 1);
    }


    /**
     * Return a new output stream for writing to the content. Any existing content will be overwritten.
     * @return
     */
    std::ostream& Content::getOutputStream() {
      throw string("operation not yet supported!");
    }

    /**
     *  Return a new input stream for reading the content.
     * @return
     */
    std::istream& Content::getInputStream() {
      throw string("operation not yet supported!");
    }

  }
}
